wcfsetup/install/files/ts/WoltLabSuite/Core/Bbcode/Code.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * Highlights code in the Code bbcode.
+ *
+ * @author	Tim Duesterhus
+ * @copyright	2001-2019 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module	WoltLabSuite/Core/Bbcode/Code
+ */
+
+import * as Language from "../Language";
+import * as Clipboard from "../Clipboard";
+import * as UiNotification from "../Ui/Notification";
+import Prism from "../Prism";
+import * as PrismHelper from "../Prism/Helper";
+import PrismMeta from "../prism-meta";
+
+const CHUNK_SIZE = 50;
+
+async function waitForIdle(): Promise<void> {
+  return new Promise(function (resolve, _reject) {
+    if ((window as any).requestIdleCallback) {
+      (window as any).requestIdleCallback(resolve, { timeout: 5000 });
+    } else {
+      setTimeout(resolve, 0);
+    }
+  });
+}
+
+class Code {
+  private codeContainer: HTMLElementTagNameMap["code"];
+  private language: string | undefined;
+
+  constructor(private readonly container: HTMLElement) {
+    this.container = container;
+    this.codeContainer = this.container.querySelector(".codeBoxCode > code") as HTMLElementTagNameMap["code"];
+
+    this.language = Array.from(this.codeContainer.classList)
+      .find((klass) => /^language-([a-z0-9_-]+)$/.test(klass))
+      ?.replace(/^language-/, "");
+  }
+
+  public static processAll(): void {
+    document.querySelectorAll(".codeBox:not([data-processed])").forEach((codeBox: HTMLElement) => {
+      codeBox.dataset.processed = "1";
+
+      const handle = new Code(codeBox);
+
+      if (handle.language) {
+        void handle.highlight();
+      }
+
+      handle.createCopyButton();
+    });
+  }
+
+  public createCopyButton(): void {
+    const header = this.container.querySelector(".codeBoxHeader");
+
+    if (!header) {
+      return;
+    }
+
+    const button = document.createElement("span");
+    button.className = "icon icon24 fa-files-o pointer jsTooltip";
+    button.setAttribute("title", Language.get("wcf.message.bbcode.code.copy"));
+    button.addEventListener("click", () => {
+      void Clipboard.copyElementTextToClipboard(this.codeContainer).then(() => {
+        UiNotification.show(Language.get("wcf.message.bbcode.code.copy.success"));
+      });
+    });
```

Comment:
```suggestion
    button.addEventListener("click", async () => {
      await Clipboard.copyElementTextToClipboard(this.codeContainer);

      UiNotification.show(Language.get("wcf.message.bbcode.code.copy.success"));
    });
```
This isn't NodeJS, stop marching to the right 😛

Reply:
eslint does not accept `async` functions for callbacks (because they do not return `void`).

Reply:
We could toggle `checksVoidReturn` of [`@typescript-eslint/no-misused-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md#options) to allow these uses. I see nothing wrong with this?

Reply:
Okay, I adjusted that. Misuses in event listeners should be found during review.